corehq/apps/users/util.py

```diff
@@ -327,3 +330,51 @@ def log_user_role_update(domain, user, by_user, updated_via):
             message = f"role: {user_role.name}[{user_role.get_id}]"
     message += f", updated_via: {updated_via}"
     log_model_change(by_user, user.get_django_user(), message=message)
+
+
+def log_user_change(domain, couch_user, changed_by_user, changed_via=None,
+                    message=None, fields_changed=None, action=ModelAction.UPDATE):
+    """
+    Log changes done to a user.
+    For a new user or a deleted user, log only specific fields.
+
+    :param domain: domain where the update was initiated
+    :param couch_user: user being changed
+    :param changed_by_user: user making the change
+    :param changed_via: changed via medium i.e API/Web
+    :param message: Optional Message text
+    :param fields_changed: dict of user fields that have changed with their current value
+    :param action: action on the user
+    """
+    from corehq.apps.users.models import UserHistory
+
+    # domain is essential to filter changes done in a domain
+    if not domain:
+        raise ValueError("Please pass domain")
+
+    # for an update, there should always be fields that have changed
+    if action == ModelAction.UPDATE and not fields_changed:
+        raise ValueError("'fields_changed' is required for update.")
```

Comment:
if `fields_changed` is passed for create / delete it is ignored. What do you think have having separate functions for create / update / delete:

```python
def log_user_create(domain, couch_user, changed_by_user, changed_via=None, message=None)
def log_user_update(domain, couch_user, changed_by_user, fields_changed, changed_via=None, message=None)
def log_user_delete(domain, couch_user, changed_by_user, changed_via=None, message=None)
```

Reply:
> if fields_changed is passed for create / delete it is ignored

Great point. I missed noticing that. It is not expected that `fields_changed` would be present for a create/delete, which is currently not explicit in the method. I will be happy to add that if that seems useful. I could not think of a use case where that would happen.

Having different methods would be a nice separation, though i am not sure if it adds enough value.
Here is what I had in mind with single method,
The good thing about a single method is ability to see all branches (`skip_domain` and `SYSTEM_USER_ID` submissions at one place and it's not big enough to break down yet. Create and Delete are going to be exactly the same with just the difference in action they store.

Though i like the fact that one does not need to pass the action anymore which is going to be explicit in the method name.
And, iirc, can_skip_domain is only in Create whereas SYSTEM user submission in delete, which would be nice as well to be in specific create/delete methods.

I am happy to go either way. I see both approaches have their advantages.

Reply:
The functions could still use a single private function in the background but I think it makes the usage of the function easier since you don't need to know which fields are required or not required.

Reply:
Hey @snopoke , i am going to keep this as a follow up PR if this is not a blocker.
This PR is getting pretty huge, so want to avoid changing anything till this is merged and rolled out.